A userspace C system call library for H1B peripherals.

phil-levis-google opened the pull request:
This pull request adds a userspace C library for system calls provided by H1B drivers. It adds 4 drivers:

  - aes: This is an extension of the standard AES system call interface to also include ECB mode, which is required for FIPS. Note that this is not cross-compatible with the standard AES implementation, but that's OK, since it hasn't been standardized.
  - dcrypto: access to the dcrypto big number arithmetic engine.
  - digest: access to the SHA/HMAC engine.
  - u2f: access to USB transport over endpoint 1 (EP1).

phil-levis-google committed:
Add libgolf2 and Makefile.

phil-levis-google committed:
Factor out golf2 system calls into libgolf2.

phil-levis-google committed:
Makefile for libgolf2.

phil-levis-google committed:
Makefile for libgolf2.

phil-levis-google committed:
H1B, not Hotel.

phil-levis-google committed:
Fix license and comment on Makefile.

jrvanwhy commented on userspace/libgolf2/aes_syscalls.c:
```diff
@@ -0,0 +1,204 @@
+// Struct used for creating synchronous versions of functions.
+//
+// fired -  set when the callback has been called
+// error - error received from the kernel less than zero indicates an error
```
Add a period between the sentences (after "kernel")

phil-levis-google replied:
I introduced tock_ to minimize namespace conflicts. In porting over this U2F app, I've realized that making libtock-c APIs *not* have tock_ in their names is problematic. For example, U2F defines function such as `sha256`. We can't have such functions defined by libtock.

jrvanwhy replied:
I see the need for namespacing, I'm just not sure about the namespace name. Perhaps "tock" is the right name, or perhaps "h1b" or "golf2" would be better.

phil-levis-google replied:
Fixed

phil-levis-google replied:
I think all system calls should be prefaced by tock_ since they're implemented in the kernel. But if a system call is platform or chip specific, that could be in the name as well (for clarity, not namespace reasons). E.g., `tock_h1b_keyladder`. In the case of AES, though, this API shouldn't be golf2/h1b-specific.

jrvanwhy commented on userspace/libgolf2/aes_ecb_syscalls.h:
```diff
@@ -0,0 +1,100 @@
+int tock_aes128_decrypt_cbc_sync(unsigned char* buf, unsigned char buf_len,
+                            const unsigned char* iv, unsigned char iv_len);
+
+
```
We should probably discuss this API in person. It's unclear to me how padding is handled (given how a sub-block-length buffer may be passed in), and why some values (e.g. ctr) are taken as parameters while others are separately configured (key).

phil-levis-google replied:
I agree, the comment should be clearer. It's worth noting that this is copied from `libtock-c/libtock/aes.h`; that version of the AES syscall API does not support ECB mode, which is needed for FIPS.

phil-levis-google replied:
Sub-block length operations are zero-padded.

This API should be cleaned up. `set_ctr` should not be exposed, and whether the counter is modified should be specified.

phil-levis-google replied:
Fixed.

jrvanwhy commented on userspace/libgolf2/digest_syscalls.h:
```diff
@@ -0,0 +1,49 @@
+  DIGEST_MODE_SHA256_HMAC = 2,
+} TockDigestMode;
+
+// TODO: Should be const, but currently not allowed by kernel
```
You can be const-correct in this API and non-const-correct in the syscall API, as long as the syscall driver does not actually modify the buffer.

phil-levis-google replied:
The issue is that if you pass a const buffer, it could be in RO memory (flash); the current syscall implementations only allow you to allow() buffers that are in RAM. So the const is removed to ensure you don't try passing a const buffer.

jrvanwhy replied:
Oh right, I keep forgetting about that.

jrvanwhy reviewed (changes requested):
High-level question that may need a discussion to answer: Where do we draw the line between "Tock" and the "kernel"? E.g. a lot of these function names start with "tock_" and are interfaces to functionality provided by the kernel but which are not part of the upstream Tock OS.

phil-levis-google committed:
Clean up API.
  - remove internal functions from header file
  - move #defines into implementation (since not needed by callers)
  - remove const from functions as may lead to passing RO buffers
  - specified behavior of buffers <= size of block
  - fixed some typos

phil-levis-google commented:
OK. I've gone over the comments, ready for re-review.

jrvanwhy commented on userspace/libgolf2/aes_ecb_syscalls.h:
```diff
@@ -0,0 +1,92 @@
+// cleartexts have the same ciphertext (it leaks information).
+//
+// buf      - buffer to encrypt (should be <= 16 bytes, will be zero-padded)
+// buf_len  - length of the buffer to encrypt
```
Presumably the output of the encryption operation is written back into buffer, because I don't see where else the response could be given to the caller.

In that case, what happens if buf_len < 16? In ECB mode you can't store a partial block.

phil-levis-google replied:
It is, which makes passing less than a block a useless operation; you won't get the complete ciphertext.  But the call with succeed and you'll get the first N bytes of the ciphertext returned. 

Should the API instead be that N%16 == 0?

jrvanwhy commented on userspace/libgolf2/aes_ecb_syscalls.h:
```diff
@@ -0,0 +1,92 @@
+//
+// buf      - buffer to encrypt (should be <= 16 bytes, will be zero-padded)
+// buf_len  - length of the buffer to encrypt
+// iv       - the initialization vector (must be 16 bytes)
```
Presumably the caller is responsible for forwarding the previous ciphertext through to the next block's IV? If so, that should be documented.

phil-levis-google replied:
Yes. Will do.

phil-levis-google committed:
Rewrite of AES syscall API.

phil-levis-google committed:
Redesigned AES system call interface based on @jrvanwhy comments.

phil-levis-google commented:
OK, I've updated the AES API. Now all of the encrypt/decrypt operations

  - Operate on multi-block buffers
  - Require the passed buffer is a multiple of 16 bytes (block size)
  - Return the "expected" value of IV or CTR that would be passed in the next call; for CBC encrypt, this is the same as the returned ciphertext; for CBC decrypt it's the passed ciphertext; for CTR mode it's the counter value.

It's ready for re-review.

jrvanwhy commented on userspace/libgolf2/aes_ecb_syscalls.h:
```diff
@@ -0,0 +1,114 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
```
I just realized this file's name is "aes_ecb_syscalls.h", but it contains ECB + CTR + CBC APIs.

Should it be renamed to aes_syscalls.h?

phil-levis-google replied:
The standard Tock syscall interface (not totally working) doesn't support ECB mode. I was worried about a name collision.

jrvanwhy replied:
How about aes_h1b_syscalls.h? My concern with aes_ecb_syscalls.h is that aes_ecb_syscalls indicates it *only* contains ECB syscalls, which is inaccurate.

phil-levis-google replied:
I think I'd prefer aes_syscalls_h1b.h; does that work for you? I can add comments about how it is different than the standard syscall header file.

jrvanwhy commented on userspace/libgolf2/aes_ecb_syscalls.h:
```diff
@@ -0,0 +1,114 @@
+
+// Encrypts a payload according to aes-128 counter-mode. The counter
+// stored in ctr is incremented for each block encrypted in a single call.
+// The caller is responsible for incrementing/managing the counter across
```
"The caller is responsible for incrementing/managing the counter across calls." --> This implies that the user must increment ctr between calls to tock_aes128_encrypt_ctr_sync(), but that's not true (based on the rest of the comment).

I think the sentence can be removed without hurting understanding.

phil-levis-google replied:
What that meant is that you'll need to pass the counter back; I agree, though, that it's redundant with the rest of the text I added. Removed.

phil-levis-google committed:
Fix comment.

phil-levis-google commented:
Ready for re-review. I am ambivalent about aes_syscalls vs. aes_ecb_syscalls but part of me wants to be cautious and be clear this version includes ECB.

phil-levis-google committed:
Rename AES syscall files to be clearer that they are H1B specific.

phil-levis-google committed:
Clean up warnings in AES test application.

phil-levis-google commented:
OK, I renamed aes_ecb_sycalls to h1b_aes_syscalls, to make clear it's the H1B code. These updates to the API will need to percolate into mainline Tock (where the current AES API is mildly broken).

Ready for re-review.
